src/v/kafka/security/scram_algorithm.cc

```diff
@@ -0,0 +1,205 @@
+#include "kafka/security/scram_algorithm.h"
+
+#include "utils/base64.h"
+#include "utils/to_string.h"
+#include "utils/utf8.h"
+
+#include <boost/algorithm/string.hpp>
+
+namespace kafka {
+
+// ALPHA / DIGIT / "/" / "+"
+// NOLINTNEXTLINE
+#define BASE64_CHAR "[a-zA-Z0-9/+]"
+
+// base64-4 = 4base64-char
+// base64-3 = 3base64-char "="
+// base64-2 = 2base64-char "=="
+// base64   = *base64-4 [base64-3 / base64-2]
+// NOLINTNEXTLINE
+#define BASE64                                                                 \
+    "(?:" BASE64_CHAR "{4})*(?:" BASE64_CHAR "{3}=|" BASE64_CHAR "{2}==)?"
+
+// Printable ASCII except ","
+// NOLINTNEXTLINE
+#define PRINTABLE "[\\x21-\\x2b\\x2d-\\x7e]+"
+
+// NOLINTNEXTLINE
+#define EXTENSIONS "(,[a-zA-Z]+=[\\x01-\\x2b\\x2d-\\x7f]+)*"
+
+// UTF8-char except NUL, "=", and ","
+// NOLINTNEXTLINE
+#define VALUE_SAFE_CHAR "[\\x01-\\x2b\\x2d-\\x3c\\x3e-\\x7f]"
+
+// 1*(value-safe-char / "=2C" / "=3D")
+// NOLINTNEXTLINE
+#define SASLNAME "(?:" VALUE_SAFE_CHAR "|=2C|=3D)+"
+
+// value-char = value-safe-char / "="
+// value = 1*value-char
+// NOLINTNEXTLINE
+#define VALUE "[\\x01-\\x2b\\x2d-\\x7f]+"
+
+// NOLINTNEXTLINE
+#define ALPHA "[a-zA-Z]"
+
+// clang-format off
+static constexpr const char* client_first_message_re =
+    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"
+    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*";
+
+static constexpr const char* client_final_message_re =
+    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")";
+// clang-format on
+
+std::ostream& operator<<(std::ostream& os, const scram_credential& cred) {
+    fmt::print(
+      os,
+      "salt {} server_key {} stored_key {} iterations {}",
+      cred._salt,
+      cred._server_key,
+      cred._stored_key,
+      cred._iterations);
+    return os;
+}
+
+client_first_message::client_first_message(bytes_view data) {
+    static const std::regex re(
+      client_first_message_re, std::regex::ECMAScript | std::regex::optimize);
+
+    // std::regex really likes using std::string
+    std::string message(data.begin(), data.end());
+    validate_utf8(message);
+
+    std::smatch match;
+    if (unlikely(!std::regex_match(message, match, re))) {
+        throw scram_exception(
+          fmt::format("Invalid SCRAM client first message: {}", message));
+    }
+
+    _authzid = match[1].str();
+    _username = match[3].str();
+    _nonce = match[4].str();
+
+    auto extensions = match[5].str(); // NOLINT
+    if (extensions.empty()) {
+        return;
+    }
+
+    // split on "," following the "," prefix
+    std::vector<std::string> extension_pairs;
+    boost::split(
+      extension_pairs, extensions.substr(1), [](char c) { return c == ','; });
+
+    // split pairs on first "=". the value part may also contain "="
+    for (const auto& pair : extension_pairs) {
+        auto it = std::find(pair.cbegin(), pair.cend(), '=');
+        if (unlikely(it == pair.cend())) {
```

Comment:
i think there is a new macro in vtools called vassert_with_ctx or smth  - see vlog.h

Reply:
why would we want to assert here on input over the network that's being validated?

Reply:
no `fmt_with_ctx` adds line number to the exception in case this gets nested into a `std::nested_exception` by seastar

Reply:
i see, fmt not assert.. i'm not sure that adds any value here since we'll be catching these and converting them into kafka error responses, but maybe i'm missing something? what makes this throw special compared to the countless other throws that don't use this, or maybe its just something we are trying to use more of in general?

Reply:
fixed